### PR TITLE
use ubuntu-latest for tsan CI

### LINKED ISSUE #2266: Github CI failing for tsan test
CI Link:
https://github.com/open-telemetry/opentelemetry-cpp/actions/runs/5866011830/job/15925211292?pr=2264
```
0 / 4] [Prepa] Creating source manifest for //api/test/baggage/propagation:baggage_propagator_test
[12 / 25] Compiling googlemock/src/gmock-matchers.cc; 1s linux-sandbox ... (2 actions, 1 running)
[20 / 25] Compiling googletest/src/gtest-death-test.cc; 1s linux-sandbox ... (2 actions, 1 running)
ERROR: /home/runner/work/opentelemetry-cpp/opentelemetry-cpp/api/test/baggage/propagation/BUILD:6:8: Linking api/test/baggage/propagation/baggage_propagator_test failed: (Exit 1): gcc failed: error executing command /usr/bin/gcc @bazel-out/k8-tsan-fastbuild/bin/api/test/baggage/propagation/baggage_propagator_test-2.params

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
/usr/bin/ld.gold: error: cannot open libtsan_preinit.o: No such file or directory
collect2: error: ld returned 1 exit status
INFO: Elapsed time: 48.815s, Critical Path: 5.57s
INFO: 145 processes: 127 internal, 18 linux-sandbox.
```

 Related to this issue:
https://bugs.launchpad.net/ubuntu/+source/gcc-10/+bug/2029910

Seems some change in the libtsan0 package from ubuntu 20.04, the package doesn't include  `libtsan_preinit.o` anymore. We can try to upgrade to `ubuntu-latest` if that fixes the issue.
The PR merge is blocked till this is fixed.

### PR DESCRIPTION
**Pull request by lalitb**
Fixes #2266 

## Changes

Upgrading to latest virtual env for tsan CI.

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

**Review by ThomsonTan — APPROVED** [review]

**Review by owent — APPROVED** [review]

**Comment by owent** [comment]
The report of tsan is not a problem of this PR.But I think we should use another mutex lock in `BasicCurlHttpTests::onHttpRequest` and inside the predicate function of `wait_for` in `BasicCurlHttpTests::waitForRequests`.Using `mtx_requests` here may lead to dead lock.

**Comment by ThomsonTan** [comment]
> The report of tsan is not a problem of this PR.But I think we should use another mutex lock in `BasicCurlHttpTests::onHttpRequest` and inside the predicate function of `wait_for` in `BasicCurlHttpTests::waitForRequests`.Using `mtx_requests` here may lead to dead lock.

Does this sound like a regression?

**Comment by lalitb** [comment]
> Using mtx_requests here may lead to dead lock.

@owent - Do you mean separate mutex for `onHttpRequest` and `waitForRequests`? The mutex `mtx_requests` is being used with the conditional variable `cv_got_events` to synchronize http server and client requests. I think this should be acceptable, and not a potential race condition, unless there is some issue with thread sanitizer. Not sure if this is relevant, but I found one similar bug here - https://github.com/google/sanitizers/issues/1259, reported with gcc-11.x.

File: ext/test/http/curl_http_test.cc [hunk 1]
```diff
@@ -164,7 +164,7 @@ class BasicCurlHttpTests : public ::testing::Test, public HTTP_SERVER_NS::HttpRe
 
   bool waitForRequests(unsigned timeOutSec, unsigned expected_count = 1)
   {
-    std::unique_lock<std::mutex> lk(mtx_requests);
+    std::unique_lock<std::mutex> lk(cv_mtx_requests);
     if (cv_got_events.wait_for(lk, std::chrono::milliseconds(1000 * timeOutSec), [&] {
           //
```
Comment (owent) [inline ext/test/http/curl_http_test.cc:169]: Do we need `std::unique_lock<std::mutex> lk1(mtx_requests);` here?

**Comment by owent** [comment]
> regression



> > The report of tsan is not a problem of this PR.But I think we should use another mutex lock in `BasicCurlHttpTests::onHttpRequest` and inside the predicate function of `wait_for` in `BasicCurlHttpTests::waitForRequests`.Using `mtx_requests` here may lead to dead lock.
> 
> Does this sound like a regression?

I think it's a dead lock problem and happens under this sequence.

| Thread 1                                                               | Thread 2                                             | Comment                   |
| ---------------------------------------------------------------------- | ---------------------------------------------------- | ------------------------- |
| calling `waitForRequests()`                                            |                                                      |                           |
| run `std::unique_lock<std::mutex> lk(mtx_requests);`                   |                                                      | Lock `mtx_requests`       |
|                                                                        | calling `onHttpRequest()`                            |                           |
|                                                                        | run `std::unique_lock<std::mutex> lk(mtx_requests);` | Block wait `mtx_requests` |
| run `cv_got_events.wait_for(...)` and predicate function returns false |                                                      |                           |
| Try to lock `mtx_requests`                                             |                                                      | Block wait `mtx_requests` |

The STL implementation of `std::condition_variable::wait_for` is locking the lock again so it can block the running thread, and `std::condition_variable::notify_one` or `std::condition_variable::notify_all` will unlock it and let the blocked thread continue. But in the execution sequence, there will be a deadlock.Because Thread 2 is also blocked and will not call `cv_got_events.notify_one();`.

**Comment by lalitb** [comment]
unfortunately, the problem is now happening in ubuntu 22.04 runner too:
```
ERROR: /home/runner/work/opentelemetry-cpp/opentelemetry-cpp/api/test/nostd/BUILD:4:8: Linking api/test/nostd/function_ref_test failed: (Exit 1): gcc failed: error executing command /usr/bin/gcc @bazel-out/k8-tsan-fastbuild/bin/api/test/nostd/function_ref_test-2.params

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
/usr/bin/ld.gold: error: cannot open libtsan_preinit.o: No such file or directory
```

**Comment by lalitb** [comment]
> Could we bring the tsan issue to backlog and unblock merging of other PRs?

We have it in backlog (#2266). We can force merge any PR which are failing for tsan, while we are trying to fix this.

**Comment by lalitb** [comment]
zpages tests failing for tsan. Have removed them from the tsan build, as zpages would be deprecated soon (#2217)

**Pull request merged**